File: code/jvm/turngamesfw/castlerun-module/src/main/kotlin/pt/isel/domain/Dices.kt
package pt.isel.domain

data class Dices(var dice1: Int = 0, var dice2: Int = 0) {
    fun roll() { dice1 = randomDiceValue(); dice2 = randomDiceValue() }
    fun clear() { dice1 = 0; dice2 = 0 }
    fun sum() = dice1 + dice2

    fun areEqual () = dice1 == dice2

    fun smallest() = if (dice1 < dice2) dice1 else dice2
    fun biggest() = if (dice1 > dice2) dice1 else dice2

    private fun randomDiceValue() = (1..6).random()
}

File: code/js/turngamesfw/src/games/castlerun/domain/Dices.ts
export class Dices {
    play: DicePair = new DicePair(null, null, true);
    duel: DicePair = new DicePair(null, null, false);

    constructor(play: DicePair = new DicePair(null, null, true), duel: DicePair = new DicePair(null, null, false)) {
        this.play = play;
        this.duel = duel;
    }
}

export class DicePair {
    dice1: number;
    dice2: number;
    canRoll: boolean;

    constructor(dice1: number, dice2: number, canRoll: boolean) {
        this.dice1 = dice1;
        this.dice2 = dice2;
        this.canRoll = canRoll;
    }

    roll(): void {
        if (!this.canRoll) return

        this.dice1 = this.randomDiceValue();
        this.dice2 = this.randomDiceValue();
        this.canRoll = false;
    }

    clear(): void {
        this.dice1 = null;
        this.dice2 = null;
    }

    sum(): number {
        return this.dice1 + this.dice2;
    }

    areEqual(): boolean {
        return this.dice1 === this.dice2;
    }

    smallest(): number {
        return this.dice1 < this.dice2 ? this.dice1 : this.dice2;
    }

    biggest(): number {
        return this.dice1 > this.dice2 ? this.dice1 : this.dice2;
    }

    winner(): number {
        return this.dice1 > this.dice2 ? 1 : 2;
    }

    private randomDiceValue(): number {
        return Math.floor(Math.random() * 6) + 1;
    }
}
